// shallow-render/dist/lib/tools/create-test-module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTestModule = void 0;
const common_1 = require("@angular/common");
const core_1 = require("@angular/core");
const get_ng_module_annotations_1 = require("./get-ng-module-annotations");
const mock_provider_1 = require("./mock-provider");
const ng_mock_1 = require("./ng-mock");
const type_checkers_1 = require("./type-checkers");
function createTestModule(setup, testComponents = []) {
    let mod;
    let additionalProviders = [];
    if (type_checkers_1.isModuleWithProviders(setup.testModule)) {
        additionalProviders = setup.testModule.providers || additionalProviders;
        mod = setup.testModule.ngModule;
    }
    else {
        mod = setup.testModule;
    }
    const ngModule = get_ng_module_annotations_1.getNgModuleAnnotations(mod);
    const declarations = ng_mock_1.ngMock([...ngModule.declarations, ...setup.declarations].filter(d => d !== setup.testComponentOrService), setup);
    // Test Modules cannot directly define entryComponents. To work around this,
    // we create a new module which declares/exports all entryComponents and import
    // the module into the TestModule.
    const entryComponents = [...ng_mock_1.ngMock([...ngModule.entryComponents], setup), ...setup.declarations];
    let ShallowTestModule = class ShallowTestModule {
    };
    ShallowTestModule = __decorate([
        core_1.NgModule({
            imports: [...ng_mock_1.ngMock([...ngModule.imports, ...setup.imports], setup), common_1.CommonModule],
            declarations: [...declarations, ...testComponents],
            entryComponents,
            providers: mock_provider_1.mockProvider([...setup.providers, ...ngModule.providers, ...additionalProviders], setup),
            exports: [...declarations, ...entryComponents],
            schemas: ngModule.schemas || [],
        })
    ], ShallowTestModule);
    return ShallowTestModule;
}
exports.createTestModule = createTestModule;
